File: app/legacy/WebService/controllers/batch_controller.py
# batch_controller.py
import json, os
import re
from flask import Blueprint, render_template, request, jsonify
from controllers.scheduler import log_batch_event, save_batch_response
import requests

batch_bp = Blueprint("batch", __name__, template_folder="templates")
CONFIG_PATH = "data/batch_config.json"
os.makedirs(os.path.dirname(CONFIG_PATH), exist_ok=True)


def load_configs():
    if not os.path.exists(CONFIG_PATH):
        return []
    with open(CONFIG_PATH, "r") as f:
        return json.load(f)

def save_configs(configs):
    with open(CONFIG_PATH, "w") as f:
        json.dump(configs, f, indent=2)
    print("Saving configs to:", CONFIG_PATH)


@batch_bp.route("/batch")
def batch_page():
    return render_template("batch.html")

@batch_bp.route("/batch/configs")
def get_configs():
    return jsonify(load_configs())

def next_id_id(configs):
    # Estrai tutti gli ID numerici tipo "id_005"
    ids = [int(re.search(r"id_(\d+)", id["id"]).group(1)) for id in configs]
    max_id = max(ids) if ids else 0
    return f"id_{max_id + 1:05d}"

@batch_bp.route("/batch/save", methods=["POST"])
def save_config():
    data = request.json
    configs = load_configs()
    data["id"] = next_id_id(configs)
    configs.append(data)
    save_configs(configs)
    return jsonify({"status": "saved", "id": data["id"]})

@batch_bp.route("/batch/toggle", methods=["POST"])
def toggle_config():
    data = request.json
    configs = load_configs()
    for id in configs:
        if id["id"] == data["id"]:
            id["active"] = data["active"]
            break
    save_configs(configs)
    return jsonify({"status": "updated"})

@batch_bp.route("/batch/delete", methods=["POST"])
def delete_config():
    data = request.json
    configs = load_configs()
    configs = [id for id in configs if id["id"] != data["id"]]
    save_configs(configs)
    return jsonify({"status": "deleted"})

@batch_bp.route("/batch/logs")
def get_logs():
    path = "data/batch_log.json"
    if not os.path.exists(path):
        return jsonify([])
    with open(path, "r") as f:
        return jsonify(json.load(f))

@batch_bp.route("/batch/responses")
def get_batch_responses():
    path = "data/batch_response.json"
    if not os.path.exists(path):
        return jsonify([])
    with open(path, "r") as f:
        return jsonify(json.load(f))

@batch_bp.route("/batch/run", methods=["POST"])
def run_batch():
    data = request.json
    endpoint = data.get("endpoint")
    params = data.get("params", {})
    print(endpoint)
    print(data)
    print(params)
    try:
        log_batch_event(f"Execution {id['id']} → {id['endpoint']} con {id['params']}")
        res = requests.get("http://localhost:8000" + id["endpoint"], params=id["params"])
        save_batch_response(id["id"], res.text)
        print(f"[{id['id']}] {id['endpoint']} → {res.status_code}")

        return jsonify({"status": "ok", "response": res.text})
    except Exception as e:
        return jsonify({"error": str(e)}), 500

File: BatchOrchestrator/Controllers/BatchController.cs
using Microsoft.AspNetCore.Mvc;
using BatchOrchestrator.Access;
using BatchOrchestrator.Models;
using System.Text.Json;

namespace BatchOrchestrator.Controllers
{
    [ApiController]
    [Route("batch")]
    public class BatchController : ControllerBase
    {
        private readonly IPythonApiAccess _apiAccess;

        public BatchController(IPythonApiAccess apiAccess)
        {
            _apiAccess = apiAccess;
        }

        [HttpPost("run")]
        public async Task<IActionResult> RunBatch([FromBody] BatchRequest request)
        {
            var result = await _apiAccess.CallPythonBatchAsync(request.Endpoint, request.Params, request.Id);
            return Ok(new { status = "ok", response = result });
        }

        [HttpGet("configs")]
        public IActionResult GetConfigs()
        {
            var path = "data/batch_config.json";
            if (!System.IO.File.Exists(path))
                return Ok(new List<object>());

            var json = System.IO.File.ReadAllText(path);
            return Content(json, "application/json");
        }
        [HttpGet("logs")]
        public IActionResult GetLogs()
        {
            var path = "data/batch_log.json";
            if (!System.IO.File.Exists(path))
                return Ok(new List<object>());

            var json = System.IO.File.ReadAllText(path);
            return Content(json, "application/json");
        }

        [HttpGet("responses")]
        public IActionResult GetResponses()
        {
            var path = "data/batch_response.json";
            if (!System.IO.File.Exists(path))
                return Ok(new List<object>());

            var json = System.IO.File.ReadAllText(path);
            return Content(json, "application/json");
        }

        [HttpPost("toggle")]
        public IActionResult toggle([FromBody] JsonElement request)
        {
            var id = request.GetProperty("id").GetString();
            var path = "data/batch_config.json";
            var json = System.IO.File.ReadAllText(path);
            var list = JsonSerializer.Deserialize<List<JsonElement>>(json);
            //var updated = new Dictionary<string, object>();

            var updatedList = new List<JsonElement>();

            foreach (var item in list)
            {
                var currentId = item.GetProperty("id").GetString();

                if (currentId == id)
                {
                    var updated = new Dictionary<string, object>();

                    foreach (var prop in item.EnumerateObject())
                    {
                        if (prop.Name == "active")
                        {
                            var currentValue = prop.Value.GetBoolean();
                            updated["active"] = !currentValue; // inverti
                        }
                        else
                        {
                            updated[prop.Name] = prop.Value;
                        }
                    }

                    var jsonObj = JsonSerializer.Serialize(updated);
                    var newElement = JsonSerializer.Deserialize<JsonElement>(jsonObj);
                    updatedList.Add(newElement);
                }
                else
                {
                    updatedList.Add(item); // non modificare
                }
            }
            var updatedJson = JsonSerializer.Serialize(updatedList, new JsonSerializerOptions { WriteIndented = true });
            System.IO.File.WriteAllText("data/batch_config.json", updatedJson);

            return Ok(new { id });
        }

        [HttpPost("delete")]
        public IActionResult deleteConfig([FromBody] JsonElement request)
        {
           
            if (!request.TryGetProperty("id", out var idProp))
                    return BadRequest("Missing 'id'");
            var id = idProp.GetString();
            var path = "data/batch_config.json";
            var json = System.IO.File.ReadAllText(path);
            var list = JsonSerializer.Deserialize<List<JsonElement>>(json);

            var updatedList = new List<JsonElement>();

            foreach (var item in list)
            {
                var currentId = item.GetProperty("id").GetString();

                if (currentId == id)
                {
                    var updated = new Dictionary<string, object>();

                    foreach (var prop in item.EnumerateObject())
                    {
                        updated[prop.Name] = prop.Name == "delete" ? true : prop.Value;
                    }

                    if (!updated.ContainsKey("delete"))
                        updated["delete"] = true;

                    var newJson = JsonSerializer.Serialize(updated);
                    updatedList.Add(JsonSerializer.Deserialize<JsonElement>(newJson));
                }
                else
                {
                    updatedList.Add(item);
                }
            }

            System.IO.File.WriteAllText(path, JsonSerializer.Serialize(updatedList, new JsonSerializerOptions { WriteIndented = true }));
            return Ok(new { deleted = id });



        }

        [HttpPost("save")]
        public IActionResult SaveConfig([FromBody] JsonElement data)
        {
            var path = "data/batch_config.json";
            List<JsonElement> configs;

            if (System.IO.File.Exists(path))
            {
                var existing = System.IO.File.ReadAllText(path);
                configs = JsonSerializer.Deserialize<List<JsonElement>>(existing) ?? new();
            }
            else
            {
                configs = new List<JsonElement>();
            }

            var newId = $"id_{configs.Count + 1:00000}";
            var updated = data.GetProperty("endpoint").ToString();
            var obj = JsonSerializer.Deserialize<Dictionary<string, object>>(data.GetRawText());
            obj["id"] = newId;
            configs.Add(JsonSerializer.Deserialize<JsonElement>(JsonSerializer.Serialize(obj)));

            System.IO.File.WriteAllText(path, JsonSerializer.Serialize(configs, new JsonSerializerOptions { WriteIndented = true }));
            return Ok(new { status = "saved", id = newId });
        }

        [HttpGet("healthz")]
        public async Task<IActionResult> HealthCheck()
        {
            var result = await _apiAccess.CallHealthzAsync();
            return Ok(new { status = "ok", response = result });
        }

        [HttpPost("app/rerun")]
        public async Task<IActionResult> Rerun([FromBody] JsonElement request)
        {
            var pipeline = request.GetProperty("pipeline_name_rerun").GetString();
            var bifrost = request.GetProperty("bifrost_instance").GetString();
            var result = await _apiAccess.CallRerunAsync(pipeline, bifrost);
            return Ok(result);
        }

        [HttpGet("app/runtime")]
        public async Task<IActionResult> Runtime([FromQuery] string pipeline_name, [FromQuery] string bifrost_instance)
        {
            var result = await _apiAccess.CallRuntimeAsync(pipeline_name, bifrost_instance);
            return Ok(result);
        }

        [HttpGet("app/getID")]
        public async Task<IActionResult> GetID([FromQuery] string pipeline_name, [FromQuery] string bifrost_instance, [FromQuery] string filter)
        {
            var result = await _apiAccess.CallGetIDAsync(pipeline_name, bifrost_instance, filter);
            return Ok(result);
        }

        [HttpGet("app/status")]
        public async Task<IActionResult> Status([FromQuery] string filter, [FromQuery] string bifrost_instance)
        {
            var result = await _apiAccess.CallPipelineStatusAsync(filter, bifrost_instance);
            return Ok(result);
        }

        [HttpGet("app/lastLog")]
        public async Task<IActionResult> LastLog([FromQuery] string pipeline_name, [FromQuery] string bifrost_instance)
        {
            var result = await _apiAccess.CallLastLogAsync(pipeline_name, bifrost_instance);
            return Ok(result);
        }

        [HttpGet("app/fullExtract")]
        public async Task<IActionResult> FullExtract([FromQuery] string bifrost_instance, [FromQuery] string filter)
        {
            var result = await _apiAccess.CallFullExtractAsync(bifrost_instance, filter);
            return Ok(result);
        }

        [HttpGet("app/userStatus")]
        public async Task<IActionResult> UserStatus([FromQuery] string visualfabriq_instance)
        {
            var result = await _apiAccess.CallUserStatusAsync(visualfabriq_instance);
            return Ok(result);
        }

    }
}
